### PR TITLE
Prevent memory churn described in #208

### LINKED ISSUE #208: Memory churn when collecting `CStateFlow` of `List` due to `===` test.
I have a viewModel that exposes a `CStateFlow` of `List` that I'm using in SwiftUI. It looks like this:

```kotlin
private val _items = MutableStateFlow<List<Item>>(listOf()).cMutableStateFlow()
var items: CStateFlow<List<Item>> = _items.cStateFlow()
```

In my SwiftUI view, I use the `state` helper to subscribe to the stateFlow:

```swift
struct ExampleView: View {
  @ObservableObject viewModel = ExampleViewModel()

  var body: some View {
    List(viewModel.state(\.items)) { (item: Item) in
      Text("Item: \(item.id)")
    }
  }
}
```

This setup is enough to trigger memory churn. Even though `items` never changes in the viewModel in this example, there's extremely high CPU usage and memory allocations for what should be an idle view.

I did some digging and have a general idea of the root cause. From what I can tell, asking for `stateFlow.value` inside the [`state` method](https://github.com/icerockdev/moko-mvvm/blob/develop/mvvm-flow/apple/xcode/mokoMvvmFlowSwiftUI/ViewModelState.swift#L20) allocates a new array, every time. I don't know why this happens, perhaps it has something to do with the Kotlin/Swift(ObjC) bridge. You can see this for yourself by adding some simple logging inside of the method:

```swift
var value1 = stateFlow.value
var value2 = stateFlow.value
var value3 = stateFlow.value
            
print("value1 = \(value1)")
print("value2 = \(value2)")
print("value3 = \(value3)")
print("stateFlow.value = \(stateFlow.value)")
print("stateFlow.value = \(stateFlow.value)")
print("same instances = \(stateFlow.value === stateFlow.value)")
```

This will output, in an infinite loop, something like this:

```
value1 = Optional(<SharedKListAsNSArray 0x600001987900>(

)
)
value2 = Optional(<SharedKListAsNSArray 0x6000019be500>(

)
)
value3 = Optional(<SharedKListAsNSArray 0x6000019aba80>(

)
)
stateFlow.value = Optional(<SharedKListAsNSArray 0x6000019abcc0>(

)
)
stateFlow.value = Optional(<SharedKListAsNSArray 0x6000019f76c0>(

)
)
same instances = false
```

The solution is to change the equality test of `NSArray` `state` helper at https://github.com/icerockdev/moko-mvvm/blob/develop/mvvm-flow/apple/xcode/mokoMvvmFlowSwiftUI/ViewModelState.swift#L86.

Rather than `equals: { $0 === $1 },` it should use [`isEqual(to:)`](https://developer.apple.com/documentation/foundation/nsarray/1411770-isequal) to compare array contents, `equals: { $0!.isEqual(to: $1 as! Array<T>) },`

After making this change locally, the memory churn goes away.

### PR DESCRIPTION
**Pull request by darronschall**
Fixes #208

Accessing `stateFlow.value` is not guaranteed to always return the same instance even when the underlying value is unchanged. This  stateFlow behavior was incorrectly causing infinite `objectWillChange` events due to the equality test always reporting `false` even when the new value was, in fact, equal to the previous.

The fix is to go through [`isEqual(to:)`](https://developer.apple.com/documentation/foundation/nsarray/1411770-isequal) to compare the value _contents_ to determine equality before firing off an `objectWillChange` event.

### CODE REVIEW

**Comment by Alex009** [comment]
@kramlex check this please

File: mvvm-flow/apple/xcode/mokoMvvmFlowSwiftUI/ViewModelState.swift [hunk 1]
```diff
@@ -83,7 +83,19 @@ public extension ObservableObject where Self: ViewModel {
     func state<T>(_ flowKey: KeyPath<Self, CStateFlow<NSArray>>) -> Array<T> {
         return state(
             flowKey,
-            equals: { $0 === $1 },
+            equals: { oldValue, newValue in
+                if oldValue == nil {
```
Comment (kramlex) [inline mvvm-flow/apple/xcode/mokoMvvmFlowSwiftUI/ViewModelState.swift:87]: why use a nil check and then do force unwrap, if it's easier?

Comment (kramlex) [reply mvvm-flow/apple/xcode/mokoMvvmFlowSwiftUI/ViewModelState.swift:87]:
```swift
if let oldValue = oldValue {
    guard let newValue = newValue as? Array<T> else {
        return false
    }
    return oldValue.isEqual(to: newValue)
} else {
    return newValue == nil
}
```

**Review by kramlex — APPROVED** [review]

**Comment by kramlex** [comment]
@Alex009 check this pr please

**Review by Alex009 — APPROVED** [review]

**Pull request merged**